[api-minor] Allow specifying custom match logic in PDFFindController

nicolo-ribaudo opened the pull request:
> **Allow specifying custom match logic in PDFFindController**
>
> This patch allows embedders of PDF.js to provide custom match logic for seaching in PDFs. This is done by subclassing the PDFFindController class and overriding the `match` method.
>
> `match` is called once per PDF page, receives as parameters the search query, the page contents, and the page index, and returns an array of `{ index, length }` objects representing the search results.

This is my proposed API for https://github.com/mozilla/pdf.js/issues/18482. It is mostly moving code around, to carve out a (public) method with the minimum possible API that non-Firefox embedders can use to provide their own custom search. More specifically:
- the logic in `#calculateMatch` that builds the RegExp has been moved to `#calculateRegExpMatch`, so that `#calculateMatch` is agnostic to the matching logic and only takes care of running the matcher and updating the state based on the match result
- `#calculateRegExpMatch` has been renamed to `match(...)`, that subclasses can override
- ~`#calculateMatch` supports calling `match` even when it's an async function. This does not affect PDF.js itself (since `#calculateMatch()` is already called in a non-awaited `.then(() => ...)`), but makes it possible for consumers to have async match logic.~

I believe that this API is minimal enough that it won't cause problems if in the future PDFFindController needs to be refactored, as @calixteman mentioned in https://github.com/mozilla/pdf.js/issues/18482#issuecomment-2248020959.

Some examples of how it can be used:


<details>
<summary><em>External search provider</em></summary>

```js
import fuzzySearch from "some-fuzzy-search-library";

class FuzzyFindController extends PDFFindController {
  // "query" is a string
  match(query, text) {
    const results = fuzzySearch(query, text);
    return results.map(({ index, value }) => ({ index, length: value.length }));
  }
}
```

</details>

<details>
<summary><em>Multi-word search</em></summary>

This is already supported by PDF.js, but as far as I can tell it cannot be used through the Firefox UI. This example is how it would be implemented in an alternative timeline where this PR would have happened before adding support for multi-word search.

This example assumes that in that alternative universe `convertToRegExpString` is not private, and it accepts `pageIndex` instead of `this._hasDiacritics[pageIndex]`.

```js
class MultiWordFindController extends PDFFindController {
  // "query" is an array of strings
  match(query, text, pageIndex) {
    let isUnicode = false;
    // Words are sorted in reverse order to be sure that "foobar" is matched
    // before "foo" in case the query is "foobar foo".
    const queryStr = query
      .sort()
      .reverse()
      .map(q => {
        const [isUnicodePart, queryPart] = this.convertToRegExpString(
          q,
          pageIndex
        );
        isUnicode ||= isUnicodePart;
        return `(${queryPart})`;
      })
      .join("|");

    const flags = `g${isUnicode ? "u" : ""}${this.state.caseSensitive ? "" : "i"}`;
    query = new RegExp(queryStr, flags);

    const matches = [];
    for (const { index, 0: match } of pageContent.matchAll(query)) {
      matches.push({ index, length: match.length });
    }
    return matches;
  }
}
```

</details>

<details>
<summary><em>Simple multi-page search</em></summary>

**EDIT**: This example does not apply anymore now that we only support sync `.match`. See https://github.com/mozilla/pdf.js/pull/18549#issuecomment-2268954354 for an async matcher example.

This implementation uses some `_`-prefixed properties of `PDFFindController`. Assuming that they are meant to be private (I can open a PR to replace `_` with `#` if needed, after that this PR lands), there is also a second implementation that only uses the real public API.

```js
class MultiPageFindController extends PDFFindController {
  // "query" is a string
  async match(query, text, pageIndex) {
    let prefix = "", suffix = "";
    if (pageIndex > 0) {
      await this._extractTextPromises[pageIndex - 1];
      prefix = this._pageContents[pageIndex - 1].slice(1 - query.length) + " ";
    }
    if (pageIndex + 1 < this._linkService.pagesCount) {
      await this._extractTextPromises[pageIndex + 1];
      suffix = " " + this._pageContents[pageIndex + 1].slice(0, query.length - 1);
    }
    text = prefix + text + suffix;

    const matches = [];
    let index = -1;
    while ((index = text.indexOf(query, index + 1)) !== -1) {
      let start = Math.max(prefix.length, index);
      let end = Math.min(index + query.length, prefix.length + text.length);
      matches.push({ index: start - prefix.length, length: end - start });
    }
    return matches;
  }
}
```

```js
class MultiPageFindController extends PDFFindController {
  #linkService;
  #pageContents;
  #pageContentsPromises = [];

  constructor(opts) {
    super(opts);
    this.#linkService = opts.linkService;
  }

  async #getPageContent(index) {
    if (this.#pageContents[index] == null) {
      this.#pageContentsPromises[pageIndex - 1] ??= Promise.withResolvers();
      await this.#pageContentsPromises[pageIndex - 1].promise;
    }
    return this.#pageContents[index];
  }

  // "query" is a string
  async match(query, text, pageIndex) {
    if (this.#pageContents[pageIndex] == null) {
    	this.#pageContents[pageIndex] = text;
    	this.#pageContentsPromises[pageIndex]?.resolve();
    }

    const [prevPage, nextPage] = await Promise.all([
      pageIndex > 0 ? this.#getPageContent(pageIndex - 1) : "",
      pageIndex + 1 < this.#linkService.pagesCount ? this.#getPageContent(pageIndex + 1) : "",
    ]);
    const prefix = prevPage.slice(1 - query.length) + " ";
    const suffix = " " + nextPage.slice(0, query.length - 1);
    text = prefix + text + suffix;

    const matches = [];
    let index = -1;
    while ((index = text.indexOf(query, index + 1)) !== -1) {
      let start = Math.max(prefix.length, index);
      let end = Math.min(index + query.length, prefix.length + text.length);
      matches.push({ index: start - prefix.length, length: end - start });
    }
    return matches;
  }
}
```

</details>

There are two questions that for which I keep swinging back and forth:
- Should the API be subclass-based, or parameter-based?
  ```js
  class CustomFindController extends PDFFindController {
    match(...) {}
  }
  ```
  vs
  ```js
  new PDFFindController({
    /* ... various other options... ,*/
    matcher(...) {}
  }
  ```
  `PDFFindController` already accepts multiple parameters to control its behavior, but having it as a subclass extension point leads to a cleaner implementation.
- Should the `isEntireWord` word check apply after running the matcher, or as part of the default matcher? Running after means that it would easily be available to custom matching logic (simply by setting `eintireWord: true` on the dispatched `"find"` event), but on the other hand it feels like its part of the match logic itself.

Please let me know what you think about this :)

PS. If this feature gets accepted and it will need any maintenance in the future feel free to ping me, similarly to how I have been keeping an eye on the various bugs related to text selection.

nicolo-ribaudo committed:
Allow specifying custom match logic in PDFFindController

This patch allows embedders of PDF.js to provide custom match
logic for seaching in PDFs. This is done by subclassing the
PDFFindController class and overriding the `match` method.

`match` is called once per PDF page, receives as parameters the
search query, the page contents, and the page index, and returns
an array of { index, length } objects representing the search
results.

Snuffleupagus commented on web/pdf_find_controller.js:
```diff
@@ -771,13 +740,65 @@ class PDFFindController {
 
-  #calculateMatch(pageIndex) {
-    let query = this.#query;
+  async #calculateMatch(pageIndex) {
```
Sorry, but why make this asynchronous?

Snuffleupagus commented on web/pdf_find_controller.js:
```diff
@@ -771,13 +740,65 @@ class PDFFindController {
     }
-    const { caseSensitive, entireWord } = this.#state;
     const pageContent = this._pageContents[pageIndex];
+    const matcherResult = await this.match(query, pageContent, pageIndex);
```
That method isn't async, so I really don't understand why `await` is necessary here!?

As far as I'm concerned this adds, completely undesirable, maintenance overhead since looking at the other code it's impossible to understand why this is done.

Snuffleupagus commented on web/pdf_find_controller.js:
```diff
@@ -799,34 +820,22 @@ class PDFFindController {
         })
         .join("|");
     }
+    if (!query) {
+      return undefined;
+    }
```
This seems to have lost the comment from https://github.com/mozilla/pdf.js/pull/18549/files#diff-693ef40ef030840f29fc58d0ec7dc7111114f096a6962d131aac3df4186e4430L677-L678

Snuffleupagus commented on web/pdf_find_controller.js:
```diff
@@ -799,34 +820,22 @@ class PDFFindController {
-      if (pageMatchesCount > 0) {
-        this.#updateUIResultsCount();
+    const matches = [];
+    for (const { index, 0: match } of pageContent.matchAll(query)) {
```
Sorry, but this looks like a completely unrelated change that we shouldn't "sneak" into *this* patch.

Snuffleupagus commented on web/pdf_find_controller.js:
```diff
@@ -930,10 +939,9 @@ class PDFFindController {
-          this._pendingFindMatches.delete(i);
-          this.#calculateMatch(i);
-        });
+        this._extractTextPromises[i]
+          .then(() => this.#calculateMatch(i))
+          .finally(() => this._pendingFindMatches.delete(i));
```
As above, this looks unrelated to *this* patch.

Snuffleupagus reviewed (changes requested):
Based on a very quick look, there appears to be some *unrelated* changes in the patch.

nicolo-ribaudo commented:
Thanks for the first review! I addressed all the comments except for those regarding the changes related to the new `await`.

The reason I added `await` in front of the `this.match` call (and thus, for making `#calculateMatch` async) is so that subclasses can easily use async search logic (mostly, for calling to external services) without being limited by the sync-ness of the API. This causes minimal changes to `PDFFindController`, since `#calculateMatch` was already called in a `.then` callback. I tried to document the need for `await` in the JSDoc comment of `match`, which lists `Promise<SingleFindMatch[]> | SingleFindMatch[]` as the return value.

While this `await` is a very-nice-to-have, it's not strictly necessary for consumers that need async search logic. Even if `PDFFindController` only supported sync searches, they would still be able to use async logic by triggering two separate `find` events with the same query (one that spawns all the search jobs and returns no matches, and then once they are done one to collect all the results) — is that `await` is a blocking concern for this PR and my explanation isn't convincing, I can remove it.

Snuffleupagus commented:
> I tried to document the need for `await` in the JSDoc comment of `match`, which lists `Promise<SingleFindMatch[]> | SingleFindMatch[]` as the return value.

Sure, but my issue is that it's completely impossible to understand that from looking only at the code *itself*. Hence why it feels like increasing the maintenance burden, since you need to either remember or (somehow) figure out why the code has unneeded asynchronicity. (Unless I'm missing things, the async-support also doesn't appear to be tested...)

> While this `await` is a very-nice-to-have, it's not strictly necessary for consumers that need async search logic.

If, and that's a very big if in my opinion, we should even consider that there needs to be actual users wanting this; not just that it'd be theoretically nice to have.
So, can we please skip the extra asynchronicity for now, and wait until an actual real-world use-cases (that cannot be solved otherwise) emerges first?

nicolo-ribaudo commented:
Well the reason I added it is that the application I'm working on would use it, it's not a theoretical use case. :stuck_out_tongue:  More specifically, we rely on an external search provider that supports searching semantically based on "similar meaning". This provider is however running asynchronously, and there is no way for me to call it synchronously.

However, as I mentioned above, I believe I can workaround a sync-only API as follows:
```js
class AsyncPDFFindController extends PDFFindController {
  #eventBus;

  #currentQuery = null;
  #pendingMatches = new Map();
  #matchResults = new Map();

  constructor(opts) {
    super(opts);
    this.#eventBus = opts.eventBus;
  }

  match(query, text, pageIndex) {
    if (this.#currentQuery !== query) {
      this.#matchResults.clear();
      this.#pendingMatches.clear();
      this.#currentQuery = query;
    }

    if (this.#matchResults.has(pageIndex)) {
      return this.#matchResults.get(pageIndex);
    }

    if (!this.#pendingMatches.has(pageIndex)) {
      this.#pendingMatches.set(
        pageIndex,
        this.matchAsync(query, text, pageIndex).then(matches => {
          this.#matchResults.set(pageIndex, matches);
          this.#pendingMatches.delete(pageIndex);
        })
      );
    }

    const { state } = this;
    if (state.type !== "custom-reloadmatches") {
      this.#pendingMatches.get(pageIndex).then(() => {
        this.#eventBus.dispatch("find", {
          ...state,
          type: "custom-reloadmatches",
        });
      });
    }

    return undefined;
  }

  async matchAsync() {
    throw new Error("Must be implemented by a sub-class");
  }
}
```

And then I can have my own async search provider by extending this `AsyncPDFFindController` class and defining a `matchAsync` method. It's not as nice as `PDFFindController` directly supporting an async match, but it should work too. For now I will remove it from this PR, and I will come back in the future if my approach ends up not working.

I agree that if we end up having async support I need to add a test for it.

timvandermeij commented:
/botio-linux preview

timvandermeij commented:
/botio unittest

timvandermeij commented:
/botio integrationtest

timvandermeij commented on web/pdf_find_controller.js:
```diff
@@ -772,12 +741,64 @@ class PDFFindController {
+  }
+
+  /**
+   * @typedef {Object} SingleFindMatch
```
I would suggest `FindMatch` instead of `SingleFindMatch` here and in the JSDoc comment below because it's shorter and "match" already indicates that it's a single one (otherwise I'd have expected "matches" or a type like `FindMatch[]` like you have below).

timvandermeij reviewed (approved):
Looks good to me, with one comment. Now that the asynchronous bits are removed and it's mainly moving existing code around I think that this refactoring is small and self-contained enough to be accepted.

Before we merge this let's await a check from @Snuffleupagus' too, given the previous questions about the implementation, to make sure we're all aligned. Thanks!

Snuffleupagus commented on web/pdf_find_controller.js:
```diff
@@ -772,12 +741,64 @@ class PDFFindController {
+    }
+
+    // Update the match count.
+    const pageMatchesCount = this._pageMatches[pageIndex].length;
```
This could/should use the local variable now.
```suggestion
    const pageMatchesCount = matches.length;
```

Snuffleupagus reviewed (approved):
r=me, with the remaining comments addressed and passing tests; thank you.

Snuffleupagus commented:
> Now that the asynchronous bits are removed and it's mainly moving existing code around I think that this refactoring is small and self-contained enough to be accepted.

Agreed; since thinking more about the suggested async behaviour of the `match`-method, I'm less convinced that it'd have been generally safe and correct unfortunately. With that being async it'd have been possible for e.g. the active search-term to change while a previously pending (and slow) `match`-call resolves, in which case we'd then update various state with "old" data.

nicolo-ribaudo commented:
Updated — thanks for the reviews!

Snuffleupagus commented:
/botio unittest

Snuffleupagus commented:
/botio integrationtest

nicolo-ribaudo commented:
Regarding the timeout on Windows for "must check the new alt text flow" and "New alt-text flow", this branch does not include those tests because I have not rebased it. Should I rebase? Or is it a flaky test?

Snuffleupagus commented:
> Regarding the timeout on Windows for "must check the new alt text flow" and "New alt-text flow", this branch does not include those tests because I have not rebased it. Should I rebase? Or is it a flaky test?

I ignored that failing test here, since I don't understand how this PR could affect that one.
Given that it's a new test I'm guessing that it's got some intermittent problems.

timvandermeij commented:
Yikes; thanks for noticing this. I have included this new one in the list of intermittents at #18396 for our overview.
